File: Shared/UsageSummary.swift
import Foundation

/// What the widget shows for a snapshot at a given moment, whatever the plan.
struct UsageSummary {
    struct Metric: Identifiable {
        var id: String
        var title: String
        var shortTitle: String?
        var value: String
        var fraction: Double
        var severity: UsageLimit.Severity
        var detail: String?
    }

    var planName: String
    var primary: Metric
    var secondary: [Metric]
    var mood: ClawdMood
    var activity: Activity?
    var note: String?
    var hasData: Bool

    init(snapshot: UsageSnapshot?, date: Date) {
        guard let snapshot else {
            planName = "Clawdmeter"
            primary = Metric(id: "empty", title: "Waiting for data", value: "Hi!", fraction: 0, severity: .normal, detail: nil)
            secondary = []
            mood = .sleeping
            activity = nil
            note = "Open the app to start"
            hasData = false
            return
        }

        planName = snapshot.plan.name
        activity = snapshot.activity
        hasData = true

        var limits = snapshot.limits.map { Self.metric(for: $0, date: date) }
        if let extra = snapshot.extraUsage, let metric = Self.metric(for: extra) {
            limits.append(metric)
        }

        if let first = limits.first {
            primary = first
            secondary = Array(limits.dropFirst())
        } else {
            primary = Self.tokenMetric(for: snapshot.activity)
            secondary = []
        }

        mood = Self.mood(for: snapshot, date: date)
        note = Self.note(for: snapshot, date: date)
    }

    private static func metric(for limit: UsageLimit, date: Date) -> Metric {
        let percent = limit.percent(at: date)
        var detail: String?
        if let resetsAt = limit.resetsAt, resetsAt > date {
            detail = Format.reset(resetsAt, from: date)
        } else if limit.resetsAt != nil {
            detail = "Fresh window"
        }
        return Metric(
            id: limit.id,
            title: limit.title,
            value: Format.percent(percent),
            fraction: percent / 100,
            severity: limit.severity(at: date),
            detail: detail
        )
    }

    private static func metric(for extra: ExtraUsage) -> Metric? {
        let used = Format.money(extra.used, currencyCode: extra.currencyCode)
        guard let percent = extra.percent, let limit = extra.limit else {
            return Metric(id: "extra", title: "Extra usage", shortTitle: "Extra", value: used, fraction: 0, severity: .normal, detail: "No monthly cap")
        }
        let severity: UsageLimit.Severity = percent >= 90 ? .critical : percent >= 75 ? .warning : .normal
        return Metric(
            id: "extra",
            title: "Extra usage",
            shortTitle: "Extra",
            value: Format.percent(percent),
            fraction: percent / 100,
            severity: severity,
            detail: "\(used) of \(Format.money(limit, currencyCode: extra.currencyCode))"
        )
    }

    private static func tokenMetric(for activity: Activity?) -> Metric {
        let today = activity?.todayTokens ?? 0
        let busiest = activity?.days.map(\.tokens).max() ?? 0
        return Metric(
            id: "tokens",
            title: "Tokens today",
            value: Format.tokens(today),
            fraction: busiest > 0 ? Double(today) / Double(busiest) : 0,
            severity: .normal,
            detail: activity.map { "\($0.todayMessages) requests" }
        )
    }

    private static func mood(for snapshot: UsageSnapshot, date: Date) -> ClawdMood {
        let sinceActive = snapshot.activity?.lastActiveAt.map { date.timeIntervalSince($0) } ?? .infinity
        let workingNow = sinceActive < 3 * 60

        guard let session = snapshot.session else {
            if workingNow { return .busy }
            return sinceActive > 30 * 60 ? .sleeping : .idle
        }

        let percent = session.percent(at: date)
        let weekly = snapshot.weekly?.percent(at: date) ?? 0
        if percent >= 100 { return .sleeping }
        if percent >= 80 || weekly >= 90 { return .tired }
        if percent >= 50 || weekly >= 75 || workingNow { return .busy }

        let windowOpen = session.resetsAt.map { $0 > date } ?? false
        if !windowOpen && sinceActive > 30 * 60 { return .sleeping }
        return .idle
    }

    private static func note(for snapshot: UsageSnapshot, date: Date) -> String? {
        switch snapshot.status {
        case .signedOut:
            return "Sign in to Claude Code"
        case .expired:
            return "Open Claude Code to reconnect"
        case .offline where date.timeIntervalSince(snapshot.updatedAt) > 15 * 60:
            return "Updated \(Format.relative(snapshot.updatedAt, from: date))"
        default:
            if date.timeIntervalSince(snapshot.updatedAt) > 45 * 60 {
                return "Updated \(Format.relative(snapshot.updatedAt, from: date))"
            }
            return nil
        }
    }
}

File: windows/Clawdmeter/Usage/UsageSummary.cs
namespace Clawdmeter.Usage;

public sealed record Metric(
    string Id,
    string Title,
    string? ShortTitle,
    string Value,
    double Fraction,
    Severity Severity,
    string? Detail);

/// What the panel shows for a snapshot at a given moment, whatever the plan.
public sealed class UsageSummary
{
    public string PlanName { get; }
    public Metric Primary { get; }
    public IReadOnlyList<Metric> Secondary { get; }
    public ClawdMood Mood { get; }
    public ActivityLog? Activity { get; }
    public string? Note { get; }
    public bool HasData { get; }

    public UsageSummary(UsageSnapshot? snapshot, DateTimeOffset date)
    {
        if (snapshot is null)
        {
            PlanName = "Clawdmeter";
            Primary = new Metric("empty", "Waiting for data", null, "Hi!", 0, Severity.Normal, null);
            Secondary = Array.Empty<Metric>();
            Mood = ClawdMood.Sleeping;
            Note = "Sign in to Claude Code to start";
            HasData = false;
            return;
        }

        PlanName = snapshot.Plan.Name;
        Activity = snapshot.Activity;
        HasData = true;

        var metrics = snapshot.Limits.Select(limit => MetricFor(limit, date)).ToList();
        if (snapshot.ExtraUsage is { } extra) metrics.Add(MetricFor(extra));

        if (metrics.Count > 0)
        {
            Primary = metrics[0];
            Secondary = metrics.Skip(1).ToList();
        }
        else
        {
            Primary = TokenMetric(snapshot.Activity);
            Secondary = Array.Empty<Metric>();
        }

        Mood = MoodFor(snapshot, date);
        Note = NoteFor(snapshot, date);
    }

    private static Metric MetricFor(UsageLimit limit, DateTimeOffset date)
    {
        var percent = limit.PercentAt(date);
        string? detail = null;
        if (limit.ResetsAt is { } reset)
        {
            detail = reset > date ? Format.Reset(reset, date) : "Fresh window";
        }
        return new Metric(limit.Id, limit.Title, null, Format.Percent(percent), percent / 100, limit.SeverityAt(date), detail);
    }

    private static Metric MetricFor(ExtraUsage extra)
    {
        var used = Format.Money(extra.Used, extra.CurrencyCode);
        if (extra.Percent is not { } percent || extra.Limit is not { } limit)
        {
            return new Metric("extra", "Extra usage", "Extra", used, 0, Severity.Normal, "No monthly cap");
        }
        var severity = percent >= 90 ? Severity.Critical : percent >= 75 ? Severity.Warning : Severity.Normal;
        return new Metric(
            "extra",
            "Extra usage",
            "Extra",
            Format.Percent(percent),
            percent / 100,
            severity,
            $"{used} of {Format.Money(limit, extra.CurrencyCode)}");
    }

    private static Metric TokenMetric(ActivityLog? activity)
    {
        var today = activity?.TodayTokens ?? 0;
        var busiest = activity?.Days.Select(day => day.Tokens).DefaultIfEmpty(0).Max() ?? 0;
        return new Metric(
            "tokens",
            "Tokens today",
            null,
            Format.Tokens(today),
            busiest > 0 ? (double)today / busiest : 0,
            Severity.Normal,
            activity is null ? null : $"{activity.TodayMessages} requests");
    }

    private static ClawdMood MoodFor(UsageSnapshot snapshot, DateTimeOffset date)
    {
        var sinceActive = snapshot.Activity?.LastActiveAt is { } active
            ? (date - active).TotalSeconds
            : double.PositiveInfinity;
        var workingNow = sinceActive < 3 * 60;

        if (snapshot.Session is not { } session)
        {
            if (workingNow) return ClawdMood.Busy;
            return sinceActive > 30 * 60 ? ClawdMood.Sleeping : ClawdMood.Idle;
        }

        var percent = session.PercentAt(date);
        var weekly = snapshot.Weekly?.PercentAt(date) ?? 0;
        if (percent >= 100) return ClawdMood.Sleeping;
        if (percent >= 80 || weekly >= 90) return ClawdMood.Tired;
        if (percent >= 50 || weekly >= 75 || workingNow) return ClawdMood.Busy;

        var windowOpen = session.ResetsAt is { } reset && reset > date;
        if (!windowOpen && sinceActive > 30 * 60) return ClawdMood.Sleeping;
        return ClawdMood.Idle;
    }

    private static string? NoteFor(UsageSnapshot snapshot, DateTimeOffset date)
    {
        switch (snapshot.Status)
        {
            case SnapshotStatus.SignedOut:
                return "Sign in to Claude Code";
            case SnapshotStatus.Expired:
                return "Open Claude Code to reconnect";
            case SnapshotStatus.Offline when (date - snapshot.UpdatedAt).TotalMinutes > 15:
                return $"Updated {Format.Relative(snapshot.UpdatedAt, date)}";
            default:
                if ((date - snapshot.UpdatedAt).TotalMinutes > 45)
                {
                    return $"Updated {Format.Relative(snapshot.UpdatedAt, date)}";
                }
                return null;
        }
    }
}
